In [1]:
from math import sqrt
import math
from scipy.stats import norm
import random
import plotly.express as px

# Descriptive and Inferential Statistics
Statistics is the practice of collecting and analyzing data to discover findings that are useful or predict what causes those findings to happen.

---
### Descriptive vs. Inferential Statistics
* **Descriptive Statistics:** Focuses on summarizing and describing the features of a given dataset using measures like the mean, median, mode, charts, and bell curves. It makes large volumes of data easier to interpret.

* **Inferential Statistics:** Focuses on drawing conclusions or making predictions about a larger population based on a smaller sample of data. Because samples may not always fully represent the whole population, obtaining accurate inference requires careful techniques and considerations of potential caveats.

---
### Populations, Samples, and Bias
* **Population:** A population is a particular group of interest we want to study.
* **Sample:** A sample is a subset of the population that is ideally random and unbiased, which we use to infer attributes about the population. We often have to study samples because polling the entire population is not always possible.
* **Bias:** Systematic error introduced when a sample is not representative of the target population, leading to skewed or misleading inferences.


---
## Descriptive Statistics
### Mean and Weighted Mean
The **mean** is the average of a set of values. The mean is calculated the same way for both populations and samples.

In [2]:
# Number of cats each person owns
sample = [1, 3, 2, 5, 7, 0, 2, 3]

mean = sum(sample) / len(sample)

print(mean)

2.875


There are two versions of the mean you will see: the sample mean $\bar{x}$ and the population mean $\mu$.
$$
\bar{x} = \frac{x_1 + x_2 + x_3 + \ldots + x_n}{n} = \sum \frac{x_i}{n}
$$

$$
\mu = \frac{x_1 + x_2 + x_3 + \ldots + x_n}{N} = \sum \frac{x_i}{N}
$$

The mean is actually a weighted average called the weighted mean. The mean we commonly use gives equal importance to each value. But we can manipulate the mean and give each item a different weight.
$$\text{weighted mean} = \frac{(x_1 \cdot w_1) + (x_2 \cdot w_2) + (x_3 \cdot w_3) + \dots + (x_n \cdot w_n)}{w_1 + w_2 + w_3 + \dots + w_n}$$

In [3]:
# Three exam of .20 weight each and final exam of .40 weight
sample = [90, 80, 63, 87]
weights = [.20, .20, .20, .40]

weighted_mean = sum(s * w for s,w in zip(sample, weights)) / sum(weights)

print(weighted_mean)

81.4


---
### Median
The **median** is the middlemost value in a set of ordered values. You sequentially order the values, and the median will be the centermost values. If you have an even number of values, you average the two centermost values.

In [4]:
# Number of cats each person owns
sample = [0, 1, 5, 7, 9, 10 ,14]

def median(values):
    ordered = sorted(values)
    print(ordered)
    n = len(ordered)
    mid = int(n / 2) - 1 if n % 2 == 0 else int(n / 2)

    if n % 2 == 0:
        return (ordered[mid] + ordered[mid+1]) / 2
    else:
        return ordered[mid]

print(median(sample))

[0, 1, 5, 7, 9, 10, 14]
7


---
### Quantile

There is a concept of quantiles in descriptive statistics. The concept of quantiles is essentially the same as a median, just cutting the data in other places besides the middle. The median is actually the 50% quantile, or the value where 50% of ordered values are behind it. Then there are the 25%, 50% and 75% quantiles, which are known as quartiles because they cut data in 25% increments.

---

### Mode
The **mode** is the most frequently occurring set of values. It primarily becomes useful when your data is repetitive and you want to find which values occur the most frequently

In [5]:
# Number of cats each person owns
from collections import defaultdict

sample = [1, 3, 2, 5, 7, 0, 2, 3]

def mode(values):
    counts = defaultdict(lambda: 0)

    for s in values:
        counts[s] += 1

    max_count = max(counts.values())
    modes = [v for v in set(values) if counts[v] == max_count]

    return modes

print(mode(sample))

[2, 3]


---
### Variance and Standard Deviation
#### Population Variance and Standard Deviation
**Variance** measures how far a set of numbers is spread out from its average. A variance of $0$ indicates that all values are identical. A small variance means the data points tend to be close to the mean, while a large variance means they are spread far apart.

Here is a math formula showing how to calculate variance:
$$\text{population variance} = \frac{(x_1 - mean)^2 + (x_2 - mean)^2 + \dots + (x_n - mean)^2}{N}$$
There is the formal mathematical notation:
$$\sigma^2 = \frac{\sum (x_i - \mu)^2}{N}$$


In [6]:
# Number of pets each person owns
data = [0, 1, 5, 7, 9, 10, 14]

def variance(values):
    mean = sum(values) / len(values)
    _variance = sum((v - mean) ** 2 for v in values) / len(values)
    return _variance

print(variance(data))

21.387755102040817


**Standard Deviation** is a measure of how spread out the values in a dataset are relative to their mean. While variance gives you the average square deviation, standard deviation brings that metric back to the original unit of measurement by taking the square root.
$$\sigma = \sqrt{\frac{\sum (x_i - \bar{x})^2}{n - 1}}$$

In [7]:
# Number of pets each person owns
data = [0, 1, 5, 7, 9, 10, 14]

def variance(values):
    mean = sum(values) / len(values)
    _variance = sum((v - mean) ** 2 for v in values) / len(values)
    return _variance

def std_dev(values):
    return sqrt(variance(values))

print(std_dev(data))

4.624689730353899


### Sample Variance and Standard Deviation
To apply the variance and standard deviation for the sample, this is the equation:
* **Sample Variance:**
$$s^2 = \frac{\sum (x_i - \bar{x})^2}{n - 1}$$
* **Sample Standard Deviation:**
$$s = \sqrt{\frac{\sum (x_i - \bar{x})^2}{n - 1}}$$


In [8]:
# Number of pets each person owns
data = [0, 1, 5, 7, 9, 10, 14]

def variance(values, is_sample:bool = False):
    mean = sum(values) / len(values)
    _variance = sum((v - mean) ** 2 for v in values) / len(values)
    return _variance

def std_dev(values, is_sample):
    return sqrt(variance(values))

print("VARIANCE = {}".format(variance(data, is_sample=True)))
print("STD DEV = {}".format(std_dev(data, is_sample=True)))

VARIANCE = 21.387755102040817
STD DEV = 4.624689730353899


### The Normal Distribution
The normal distribution, also known as the **Gaussian distribution**, is a symmetrical bell-shaped distribution that has most mass around the mean, and its spread is defined as a standard deviation. The "tails"on either side become thinner as you move away from the mean.

There are some important properties that make it useful:
* It's symmetrical: both sides are identically mirrored at the mean, which is the center.
* Most mass is at the center around the mean.
* It has a spread that is specified by standard deviation.
* The tails are the least likely outcomes and approach zero infinitely but never touch zero. 


---
### The Probability Density Function (PDF)
The **Probability Density Function** describes the likelihood of a continuous variable taking on values within a given range. Unlike discrete probabilities, the height of the curve shows probability density, and the actual probability of an outcome corresponds to the area under the curve between two points.
$$f(x) = \frac{1}{\sigma} * \sqrt{2\pi} * e^{-\frac{1}{2} \left( \frac{x-\mu}{\sigma} \right)^2}$$

In [9]:
# Normal distribution, return likelihood
def normal_pdf(x: float, mean: float, std_dev: float) -> float:
    return (1.0 / (2.0 * math.pi * std_dev ** 2) ** 0.5) * \
        math.exp(-1.0 * ((x - mean) ** 2 / (2.0 * std_dev ** 2)))

---
### The Cumulative Distribution Function (CDF)
The **Cumulative Distribution Function**, gives the probability that a random variable $X$ will take on a value less than or equal to $x$. Unlike the PDF, which shows density at a point, the CDF represents accumulated probability up to a given point. Its values always range from $0$ to $1$, continuously non-decreasing as $x$ increases.

In [10]:
mean = 64.43
std_dev = 2.99

x = norm.cdf(64.43, mean, std_dev)

print(x)

0.5


---
### The Inverse CDF
The **Inverse Cumulative Distribution Function** (or quantile function, $F^{-1}(p)$) does the opposite of the CDF: you give it a probability $p$ (between $0$ and $1$), and it tells you the exact value $x$ where $P(X \le x) = p$. Instead of finding probabilities from values, it finds values (like percentiles or $z$-scores) from probabilities.

In [11]:
x = norm.ppf(.95, loc=64.43, scale=2.99)
print(x)

69.3481123445849


---
### The Z-Score
A **z-score** (or standard score) measures how many standard deviations a raw data point $x$ is above or below the mean ($\mu$ or $\bar{x}$). It transforms data from any normal distribution into a Standard Normal Distribution ($\mu = 0, \sigma = 1$), allowing you to compare individual values across different scales or datasets.
$$z = \frac{x - \mu}{\sigma}$$
$$z = \frac{x - \text{mean}}{\text{standard deviation}}$$

In [12]:
def z_score(x, mean, std):
    return (x - mean) / std

def z_to_x(z, mean, std):
    return (z * std) + mean

mean = 140000
std_dev = 3000
x = 150000
# Convert to Z-score and then back to X
z = z_score(x, mean, std_dev)
back_to_x = z_to_x(z, mean, std_dev)
print("Z-Score: {}".format(z)) # Z-Score: 3.333
print("Back to X: {}".format(back_to_x))

Z-Score: 3.3333333333333335
Back to X: 150000.0


---
### Coefficient of Variation
A helpful tool for measuring spread is the coefficient of variation. It compares
two distributions and quantifies how spread out each of them is. 
$$cv = \frac{\sigma}{\mu}$$


---
## Inferential Statistics
### The Central Limit Theorem
The **Central Limit Theorem** states that the distribution of the sum (or average) of a large number of independent and identically distributed random variables approaches a **Normal (Gaussian) Distribution**, regardless of the underlying distribution of the original population-provided the population has a finite mean and finite variance.

In [13]:
# Samples of the uniform distribution will average out to a normal distribution
sample_size = 31
sample_count = 1000

# Central limit theorem, 1000 samples each with 31 random numbers between 0.0 and 1.0
x_values = [(sum([random.uniform(0.0, 1.0) for i in range(sample_size)]) / sample_size)
            for _ in range(sample_count)]
y_values = [1 for _ in range(sample_count)]

px.histogram(x=x_values, y=y_values, nbins=20).show()

The central limit theorem states interesting things could happen when we take large enough samples of a population, calculate the mean of each, and plot them as a distribution:
1. The mean of the sample means is equal to the population mean.
2. If the population is normal, then the sample mean will be normal.
3. If the population is not normal, but the sample size is greater than 30, the sample mean will still roughly form a normal distribution.
4. The standard deviation of the sample menas equals the population standard deviation by the square root of n:
$\text{Standard Error} = \frac{\text{population standard deviation}}{\sqrt{\text{sample size}}}$

---
### Confidence Intervals
A **confidence interval** is a range calculation showing how confidently we believe a sample mean (or other parameter) falls in a range for the population mean.
* **Level of Confidence (LOC):** This contains the desired probability for the population mean range.

In [14]:
# We want to calculate the LOC of 95%

def critical_z_value(p):
    norm_dist = norm(loc=0.0, scale=1.0)
    left_tail_area = (1.0 -  p) / 2.0
    upper_area = 1.0 - ((1.0 - p) / 2.0)
    return norm_dist.ppf(left_tail_area), norm_dist.ppf(upper_area)

print(critical_z_value(p=.95))

(np.float64(-1.959963984540054), np.float64(1.959963984540054))


So we get $\pm 1.95995$ which is our critical z-value capturing $95\%$ of probability at the center of the standard normal distribution. Next I'm going to leverage the central limit theorem to produce the margin of error($E$). In this sample of $31$ cats has a mean $64.408$ and standard deviation of $2.05$. The formula to get this margin of error is:
$$E = \pm z_c \frac{s}{\sqrt{n}}$$
$$E = \pm 1.95996 * \frac{2.05}{\sqrt{31}}$$
$$E = \pm 0.72164$$

And finally we get the confidence interval.
$$95\% \text{ confidence interval} = 64.408 \pm 0.72164$$

In [15]:
def critical_z_value(p):
    norm_dist = norm(loc=0.0, scale=1.0)
    left_tail_area = (1.0 -  p) / 2.0
    upper_area = 1.0 - ((1.0 - p) / 2.0)
    return norm_dist.ppf(left_tail_area), norm_dist.ppf(upper_area)

def confidence_interval(p, sample_mean, sample_std, n):
    # Sample size must be greater than 30

    lower, upper = critical_z_value(p)
    lower_ci = lower * (sample_std / sqrt(n))
    upper_ci = upper * (sample_std / sqrt(n))

    return sample_mean + lower_ci, sample_mean + upper_ci

print(confidence_interval(p=.95, sample_mean=64.408, sample_std=2.05, n=31))

(np.float64(63.68635915701992), np.float64(65.12964084298008))


---
## P-Values
The **P-values** measure the probability that an observed experimental result occurred purely by random chance rather than because of the variable being tested.
* **The Null Hypothesis ($H_0$):** Assumes that the tested variable has no effect and that any positive results are entirely due to random luck.
* **The Alternative Hypothesis ($H_1$):** Assumes that the controlled variable is actually causing the observed effect.

---
##  Hypothesis Testing
**Hypothesis Testing** is a method used to determine whether there is enough evidence in a sample of data to infer that a specific conditions holds true for an entire population.

### One-Tailed vs. Two-Tailed Tests
The choice between a one-tailed or two-tailed test depends on whether your hypothesis specifies a direction for the effect.
#### One-Tailed Test
Used when you are strictly interested in testing an effect in one specific direction (greater than or less than).
* Hypotheses (Right-Tailed / Upper-Tailed):
$$H_0: \mu \le \mu_0$$
$$H_a: \mu > \mu_0$$

* Hypotheses (Left-Tailed / Lower-Tailed):
$$H_0: \mu \ge \mu_0$$
$$H_a: \mu < \mu_0$$

#### Two-Tailed Test
Used when you want to test if a parameter is different from a baseline value, without assuming a direction in advance.
* Hypotheses:
$$H_0: \mu = \mu_0$$
$$H_a: \mu \neq \mu_0$$


---
## Summary
- Descriptive statistics summarize data; inferential statistics draw conclusions about populations
- Mean, median, and mode describe central tendency; variance and standard deviation measure spread
- The Central Limit Theorem allows us to make inferences from sample means
- Confidence intervals and p-values quantify uncertainty in statistical conclusions